tests/unit/amo/components/TestUserProfile.js

```diff
@@ -0,0 +1,318 @@
+import ErrorList from 'ui/components/ErrorList';
+import LoadingText from 'ui/components/LoadingText';
+import Rating from 'ui/components/Rating';
+import UserAvatar from 'ui/components/UserAvatar';
+import { dispatchSignInActions } from 'tests/unit/amo/helpers';
+import { fakeI18n, shallowUntilTarget } from 'tests/unit/helpers';
+
+
+describe(__filename, () => {
+  function renderUserProfile({
+    i18n = fakeI18n(),
+    params = { username: 'tofumatt' },
+    store = dispatchSignInActions({
+      userProps: {
+        display_name: 'Matt MacTofu',
+        userId: 500,
+        username: 'tofumatt',
+      },
+    }).store,
+    ...props
+  // eslint-disable-next-line padded-blocks
+  } = {}) {
+
+    return shallowUntilTarget(
+      <UserProfile i18n={i18n} params={params} store={store} {...props} />,
+      UserProfileBase,
+    );
+  }
+
+  it('renders user profile page', () => {
+    const root = renderUserProfile();
+
+    expect(root.find('.UserProfile')).toHaveLength(1);
+  });
+
+  it('dispatches fetchUserAccount action if username is not found', () => {
+    const { store } = dispatchSignInActions();
+    const dispatchSpy = sinon.spy(store, 'dispatch');
+    const username = 'i-am-not-in-the-store';
+
+    const root = renderUserProfile({ params: { username }, store });
+
+    sinon.assert.calledWith(dispatchSpy, fetchUserAccount({
+      errorHandlerId: root.instance().props.errorHandler.id,
+      username,
+    }));
+  });
+
+  it('dispatches fetchUserAccount action if username param changes', () => {
+    const { store } = dispatchSignInActions({ username: 'black-panther' });
+    const dispatchSpy = sinon.spy(store, 'dispatch');
+
+    const root = renderUserProfile({
+      params: { username: 'black-panther' },
+      store,
+    });
+
+    dispatchSpy.reset();
+
+    sinon.assert.notCalled(dispatchSpy);
+
+    root.setProps({ params: { username: 'killmonger' } });
+
+    sinon.assert.calledWith(dispatchSpy, fetchUserAccount({
+      errorHandlerId: root.instance().props.errorHandler.id,
+      username: 'killmonger',
+    }));
+  });
+
+  it('does not dispatch fetchUserAccount if username does not change', () => {
+    const { store } = dispatchSignInActions({ username: 'black-panther' });
+    const dispatchSpy = sinon.spy(store, 'dispatch');
+
+    const root = renderUserProfile({
+      params: { username: 'black-panther' },
+      store,
+    });
+
+    dispatchSpy.reset();
+
+    sinon.assert.notCalled(dispatchSpy);
+
+    root.setProps({ params: { username: 'black-panther' } });
+
+    sinon.assert.notCalled(dispatchSpy);
+  });
+
+  it('renders the user avatar', () => {
+    const root = renderUserProfile();
+
+    const headerContent = root.find('.UserProfile-user-info').prop('header');
+    const header = shallow(<div>{headerContent}</div>);
+
+    expect(header.find(UserAvatar)).toHaveProp(
+      'user', getCurrentUser(root.instance().props.store.getState().users));
```

Comment:
It feels a little fragile to rely on `.props.store.getState()`. How about doing it with more setup?

```diff
diff --git a/tests/unit/amo/components/TestUserProfile.js b/tests/unit/amo/components/TestUserProfile.js
index 2291383e5..ec603f585 100644
--- a/tests/unit/amo/components/TestUserProfile.js
+++ b/tests/unit/amo/components/TestUserProfile.js
@@ -94,13 +94,17 @@ describe(__filename, () => {
   });
 
   it('renders the user avatar', () => {
-    const root = renderUserProfile();
+    const username = 'tofumatt';
+    const { state, store } = dispatchSignInActions({
+      userProps: { username },
+    });
+    const root = renderUserProfile({ store, params: { username } });
 
     const headerContent = root.find('.UserProfile-user-info').prop('header');
     const header = shallow(<div>{headerContent}</div>);
 
     expect(header.find(UserAvatar)).toHaveProp(
-      'user', getCurrentUser(root.instance().props.store.getState().users));
+      'user', getCurrentUser(state.users));
   });
 
   it('still passes user prop to avatar while loading', () => {
```

If that setup becomes too repetitive you could add a helper to sync up `params` and `user`.

I also find it helpful to see explicit setup of data for tests that are testing that same data.

Reply:
Fair enough. It's a bit fragile but I found it less work.

On the second point I totally agree though, so yeah I'll refactor this. It's handy to see all the props set in the test. 👍